L15-PHP-Interceptors.md

```diff
@@ -0,0 +1,237 @@
+or handle each request type differently, based on the needs of the interceptor.
+The structure for methods in `Grpc\Interceptor` is shown below:
+
+```php
+class Interceptor{
+  /**
+     * @param string   $method       The name of the method to call
+     * @param mixed    $argument     The argument to the method
+     * @param array    $metadata     A metadata map to send to the server(optional)
+     * @param array    $options      An array of call_options (optional)
+     * @param function $continuation Used to invoke the next interceptor.
+     *
+     * @return \Closure A function which can create a UnaryCall
+     */
+  public function interceptUnaryUnary($method, $argument, array $metadata = [], array $options = [], $continuation){}
+
+  public function interceptStreamUnary($method, array $metadata = [], array $options = [], $continuation){}
+
+  public function interceptUnaryStream($method, $argument,array $metadata = [], array $options = [], $continuation){}
+
+  public function interceptStreamStream($method,array $metadata = [], array $options = [], $continuation){}
+}
+```
+
+The last parameter `$continuation()` is a callable value which lets the user to decide
+to invoke the next Interceptor or stop RPC. By default each method will invoke the next
+interceptor and the return value will be the Call object from the next wrapper level.
+The user can stop the RPC by not calling the `$continuation()` and return anything like \Exception or no return.
+
+As you can see from the `Grpc\Interceptor`, stream client doesn’t have the request
+argument. The user need to wrap the return value `$continuation()` which is a Call
+object into a new Call object to intercept the request. The way to intercept a
+request looks like below:
+
+```php
+Class RequestInterceptCall(
+	private $call;
+	public construct($call){
+		$this->call = $call;
+	}
+	public function write($request){
+		anything_intercept_request($request);
+		return $this->call->write($request);
+  }
+  Public function wait(){
+    return $this->call->wait();
+  }	
+)
+class UserInterceptor extends Grpc\Interceptor{
+  public function StreamUnary($method, array $metadata = [], array $options = [], $continuation){
+    return new RequestInterceptCall($continuation($method, $argument, $metadata, $options));
+  }
+  public function StreamStream($method, array $metadata = [], array $options = [], $continuation){
+    return new RequestInterceptCall($continuation($method, $argument, $metadata, $options));
+  }
+}
+```
+
+The Call intercept will be executed with the order same as the `Grpc\Interceptor`.
+In other words, if there are multiple Call intercept followed by multiple
+`Grpc\Interceptor`, the order of execution `$call->wait($request)` will start from
+the most outside wrapped call to the next level, until it hits the deepest Call object.
+
+### Client Interceptors
+#### Illustrating Example 1 (One Interceptor writing log)
```

Comment:
I suggest instead of having "test case" examples here, portraying them as high level use-cases by titling them properly. e.g. "Logging Interceptor", "Adding and Modifying Request Headers", etc.
(for the two interceptor case, find an appropriate use case?).